### PR TITLE
GH-1186: Remove Packet::hdr_size

### LINKED ISSUE #1186: Packet analysis: Calculate packet header size as each analyzer processes data
Currently the packet header size is calculated one time after all of the headers have been processed by the packet analyzers. This seems like it should be a function of the individual analyzers. They know how much data they're processing, and can just add that length to the header size as they go.

### PR DESCRIPTION
**Pull request by timwoj**
Fixes #1186

### CODE REVIEW

**Comment by timwoj** [comment]
@J-Gras You might want to look this one over

**Comment by J-Gras** [comment]
The patch looks fine to me. However, I am still a little bit hesitant about the approach. Currently, none of the existing analyzers except IP makes use of this value, right? Although we don't need the temporary values of `hdr_size`, we split the calculation. Thus we introduce more points where we can screw up the calculation. Also, the following looks like containing a lot redundancy:
```
packet->hdr_size += hdr_size;
return ForwardPacket(len - hdr_size, data + hdr_size, packet, protocol);
```
Do we really need `hdr_size` in `Packet`? If yes, we should keep in mind to update the docs, because correctly incrementing the value is crucial now.

**Comment by timwoj** [comment]
How about having `ForwardPacket` update it every time? That way the analyzers don't need to remember to do it, and it's always kept up to do date.

**Comment by J-Gras** [comment]
> How about having `ForwardPacket` update it every time? That way the analyzers don't need to remember to do it, and it's always kept up to do date.

That would help in some cases. In case the analyzer does not use `ForwardPacket()`, one would still have to do it manually. This would happen if the analyzer is the last one and the packet shouldn't be processed by the session logic. This case shouldn't be an issue as nothing would depend on the size to be set. The other situation would be that the next analyzer isn't identified using an ID (e.g. in Ethernet). Overall things might become confusing.

Another option would be to update the header size in `AnalyzePacket()`. This would require analyzers to call the base class version of the method first.

Actually the meaning of `hdr_size` isn't quite clear anyway. What about removing that field and just store a pointer to the beginning of the (first?) IP header in `Packet`? That seems to align well with the suggestions from #1185.

**Comment by rsmmr** [comment]
> Actually the meaning of `hdr_size` isn't quite clear anyway. What about removing that field and just store a pointer to the beginning of the (first?) IP header in `Packet`? That seems to align well with the suggestions from #1185.

Yeah, I think that would be ok as well.

**Comment by timwoj** [comment]
I'm intentionally leaving this one open until the IP work lands, since it makes a lot of changes to how the header size is used. It should be possible to remove it after that lands.

**Comment by timwoj** [comment]
> Actually the meaning of hdr_size isn't quite clear anyway. What about removing that field and just store a pointer to the beginning of the (first?) IP header in Packet? That seems to align well with the suggestions from #1185.

I feel like this replaces the header size with the same data in a different format. We're already storing the processed IP header in the packet object. In the places where we were using the header size solely to re-parse that header, we can just access that pointer directly. There are other places it gets used though, such as in the VXLAN analyzer, where it uses the header size to know where the data starts for the encapsulated packet to be processed.

Maybe we don't need to update it with every analyzer like I was originally planning, but I'm not so sure that storing off a pointer to the end of the headers is the right way.

**Comment by timwoj** [comment]
@J-Gras @rsmmr Thoughts on the above?

**Comment by J-Gras** [comment]
I took a look at [VXLAN](https://github.com/zeek/zeek/blob/master/src/analyzer/protocol/vxlan/VXLAN.cc). If I got that right, the "processing flow" will continue in packet analysis with the new packet. The updated `data` and `len` values are only used to provide some context for VXLAN's `ProtocolViolation`. Similar details may be provided by the IP packet analyzer for the inner packet. In that case we might omit that at this point.
From my perspective the only kind of situation where we would need to maintain an additional payload pointer is the following: There is a packet with three layers A, B and C and we have corresponding packet analyzers. Analyzer A does some sort of processing and determines that there is a layer B followed by layer C. Thus, analyzer A would call analyzer B, B does some processing but does not know how to proceed. We would come back to A, which could use `hdr_size` to skip over B and continue processing by triggering C. However, I haven't encountered that situation so far.
Looking at the reworked "processing flow", I wonder whether we still need `l2_valid` in Packet. If not, returning the number of bytes processed might be viable again for `AnalyzePacket()`. In the above example, B would return how many data it processed. A would use that value to jump to layer C and return the sum of B, C and the bytes A processed itself.

File: src/Sessions.cc [hunk 1]
```diff
@@ -81,11 +81,14 @@ void NetSessions::NextPacket(double t, Packet* pkt)
...
 
-void NetSessions::DoNextPacket(double t, const Packet* pkt)
+void NetSessions::DoNextPacket(double t, const Packet* pkt, size_t remaining)
 	{
 	const std::unique_ptr<IP_Hdr>& ip_hdr = pkt->ip_hdr;
 
-	uint32_t caplen = pkt->cap_len - pkt->hdr_size;
+	// Caplen is the amount of bytes left in the packet after all of the
+	// headers are removed.
+	uint32_t caplen = remaining;
```
Comment (J-Gras) [inline src/Sessions.cc:90]: Wouldn't it make sense to remove `caplen` and just use `remaining` in the following? I also associate `caplen` with captured length with might differ from actual packet length. To me semantics seem mixed up here.

Comment (J-Gras) [inline src/Sessions.cc:90]: Wouldn't it make sense to remove `caplen` and just use `remaining` (or `len`) in the following? I also associate `caplen` with captured length with might differ from actual packet length. To me semantics seem mixed up here.

Comment (timwoj) [reply src/Sessions.cc:90]: Yah, that confused me greatly at first too. It was just a bad variable name to begin with, hence the comment. I'll replace it throughout.

File: src/analyzer/protocol/vxlan/VXLAN.cc [hunk 2]
```diff
@@ -75,25 +75,16 @@ void VXLAN_Analyzer::DeliverPacket(int len, const u_char* data, bool orig,
 	Packet pkt(DLT_EN10MB, &ts, caplen, len, data);
 	pkt.encap = outer;
 
-	packet_mgr->ProcessPacket(&pkt);
-
-	if ( ! pkt.l2_valid )
+	if ( ! packet_mgr->ProcessInnerPacket(&pkt) )
```
Comment (rsmmr) [inline src/analyzer/protocol/vxlan/VXLAN.cc:78]: Can `bool l2_valid;` now be removed from `Packet.h`?

File: src/Sessions.h [hunk 3]
```diff
@@ -93,8 +93,13 @@ class NetSessions {
...
 	 * method is called by the packet analysis manager when after it has processed
 	 * an IP-based packet, and shouldn't be called directly from other places.
+	 *
+	 * @param t The timestamp for this packet.
+	 * @param pkt The packet being processed.
+	 * @param len The number of bytes that haven't been processed yet by packet
+	 * analysis.
 	 */
-	void DoNextPacket(double t, const Packet *pkt);
+	void DoNextPacket(double t, const Packet *pkt, size_t len);
```
Comment (rsmmr) [inline src/Sessions.h:102]: If I'm not missing anything, we now have exactly one place only that is calling `DoNextPacket`, and that's the IP analyzer forwarding TCP/UDP/ICMP. With that, it seems we can skip some of the re-processing of IP at the beginning of that method by just passing in the protocol directly. How about we change this method to `ProcessTransportLayer(double t, size_t len, const uint8_t* data, Packet* packet, int proto)` and then start directly with the transport-layer logic.

**Review by rsmmr — APPROVED** [review]
Looks good, just a couple notes.

File: src/Sessions.h (hunk 3 above)
Comment (timwoj) [reply src/Sessions.h:102]: I think it's a little weird to pass the packet object but then also pass a bunch of fields that are contained in the packet object. This code is all going to get reworked with the TCP/UDP work after the 4.0 release, so I'm a little inclined to just leave it. At the very most, I can see removing that header length check from DoNextPacket, but that's it.

File: src/analyzer/protocol/vxlan/VXLAN.cc (hunk 2 above)
Comment (timwoj) [reply src/analyzer/protocol/vxlan/VXLAN.cc:78]: Yep, removed.

File: src/Sessions.h (hunk 3 above)
Comment (rsmmr) [reply src/Sessions.h:102]: Want to rename the method at least to better express what it's doing?

**Pull request merged**